### DISCUSSION TITLE
CLEAR_EXCLUDE or PRESERVE or ?? to keep manually added folders to gh-pages.

### DISCUSSION THREAD

**Discussion by manzt** [JamesIves/github-pages-deploy-action#520] [Q&A] [answered]
Following up on the discussion here: https://github.com/JamesIves/github-pages-deploy-action/issues/202

Is there a way to specify that this action should preserve a folder from the previous `gh-pages` history?

I want to use this action to track changes in `master`, keeping my site `manzt.github.io/vizarr` up to date with the latest changes. But, I also want to preserve versioned builds of my site in a nested path (e.g. `manzt.github.io/vizarr/v0.1`), so that there are stable links that can be shared indefinitely. I've manually added the folder `v0.1` to my `gh-pages` branch, but when this action runs `v0.1` is eliminated.

Is there a way to declare folder(s)/file(s) that should remain on `gh-pages` after this action runs? Essentially I want to:

1.) Build my static site and copy the contents of `build` to root of `gh-pages` branch.
2.) Keep any folders from previous history on `gh-pages` branch that I've white-listed.

Many thanks in advance.

**Comment by Pike** [comment]
I think what happens is that the deploy on https://github.com/manzt/vizarr/commits/gh-pages that removes the versioned doc isn't for the main revision that has the exclusion. The run that has that ends up not changing the gh-pages branch, and thus doesn't push.

Can you back out https://github.com/manzt/vizarr/commit/deecebaa45ce464a0ff6ae9ce93e0a9f6da2c6de and try again?

You shouldn't need `PRESERVE` at all I think. Also, likely gonna remove that towards 4.0

**Reply by manzt** [reply]
Thanks for the reply, but I'm not sure that I understand. I manually added `v0.1/` in manzt/vizarr@deeceba to the `gh-pages` branch, and then I configured this action to commit to `gh-pages` on merge master. I want the `v0.1/` folder to remain unchanged on the `gh-pages` branch, but to have a new build replace the root everytime I merge master.

> Can you back out `manzt/vizarr@deeceba` and try again?

What is meant by "back out" here? Thanks for the help!

**Comment by Pike** [comment, marked as answer]
So, the deploy that removed v0.1 is from https://github.com/manzt/vizarr/commit/3e0871d5aa9797de248a90fcdc30ef80b0aa04e6 (you can tell from the commit message). You added the `CLEAN_EXCLUDE` in https://github.com/manzt/vizarr/commit/4b956596596cc4a1d94e1d9e83ed7b9a6f88719a, which is the next commit. That ran the publishing action, but there were no local changes, so it didn't update anything.

The removal run was probably https://github.com/manzt/vizarr/runs/1488410747, which had `CLEAN`, but not `CLEAN_EXCLUDE`. Not exactly sure where that's coming from, tbh.

You'll need to add `v0.1` again, either by doing the same steps you did the first time, or via https://git-scm.com/docs/git-restore.

**Comment by manzt** [comment]
Thanks for walking me through that! Seems to be working now. I really appreciate it.
